recirq/optimize/mpg.py

```diff
@@ -0,0 +1,390 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, List, Optional, Tuple
+
+import numpy as np
+import scipy
+from scipy.optimize import OptimizeResult
+from scipy.optimize.optimize import wrap_function
+from sklearn.linear_model import LinearRegression
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import PolynomialFeatures
+
+
+def _get_quadratic_model(
+    xs: List[np.ndarray], ys: List[float], xopt: np.ndarray
+) -> Pipeline:
+    """Fit a least squares quadratic model.
+
+    Given a list of points and their function values, fits a least squares
+    quadratic model to the function using polynomial features. Then,
+    return the model itself.
+    This function is used by `model_policy_gradient` to estimate the gradient
+    of the objective function.
+
+    Args:
+        xs: The points at which function values are given.
+        ys: The function values at the points given by `xs`.
+        xopt: The point at which to center.
+
+    Returns:
+        the regression model.
+    """
+    linear_model = LinearRegression(fit_intercept=False)
+    model = Pipeline(
+        [("poly", PolynomialFeatures(degree=2)), ("linear_model", linear_model),]
+    )
+    shifted_xs = [(x - xopt) for x in xs]
+    model = model.fit(shifted_xs, ys)
+    return model
+
+
+def clip_by_norm(t, clip_norm):
+    """Clips tensor values to a maximum L2-norm.
+    returns a tensor of the same type and shape as `t` with its values set to:
+    `t * clip_norm / l2norm(t)`
+
+    Args:
+        t: The tensor values at which we want to clip the norm.
+        clip_norm: The scalar to give a maximal upper bound for the L2-norm of the tensor.
+
+    Returns:
+        the tensor with the same shape of the input, but with norm clipped.
+    """
+    l2_norm = np.linalg.norm(t, 2)
+    return t * clip_norm / np.maximum(l2_norm, clip_norm)
+
+
+class ConstantSchedule(object):
```

Comment:
nit: don't need `object`.

Consider making this a `dataclass` with `frozen=True`, especially if you're going to use it as a default argument for a function (default arguments should not be mutable)